brainiak/matnormal/covs.py

```diff
@@ -3,438 +3,475 @@
-           'CovUnconstrainedInvCholesky',
-           'CovKroneckerFactored']
-
-
-class CovBase:
-    """Base metaclass for noise covariances
+import tensorflow_probability as tfp
+
+from brainiak.matnormal.utils import x_tx, xx_t
+from brainiak.utils.kronecker_solvers import (
+    tf_solve_lower_triangular_kron,
+    tf_solve_upper_triangular_kron,
+    tf_solve_lower_triangular_masked_kron,
+    tf_solve_upper_triangular_masked_kron,
+)
+
+__all__ = [
+    "CovBase",
+    "CovIdentity",
+    "CovAR1",
+    "CovIsotropic",
+    "CovDiagonal",
+    "CovDiagonalGammaPrior",
+    "CovUnconstrainedCholesky",
+    "CovUnconstrainedCholeskyWishartReg",
+    "CovUnconstrainedInvCholesky",
+    "CovKroneckerFactored",
+]
+
+
+class CovBase(abc.ABC):
+    """Base metaclass for residual covariances.
+    For more on abstract classes, see
+    https://docs.python.org/3/library/abc.html
+
+    Parameters
+    ----------
+
+    size: int
+        The size of the covariance matrix.
+
     """
-    __metaclass__ = abc.ABCMeta
 
     def __init__(self, size):
         self.size = size
 
+        # Log-likelihood of this covariance (useful for regularization)
+        self.logp = tf.constant(0, dtype=tf.float64)
+
     @abc.abstractmethod
     def get_optimize_vars(self):
         """ Returns a list of tf variables that need to get optimized to fit
             this covariance
         """
         pass
 
-    @abc.abstractproperty
+    @property
     def logdet(self):
-        """ log|Sigma|
+        """ log determinant of this covariance
         """
         pass
 
     @abc.abstractmethod
-    def Sigma_inv_x(self, X):
-        """Given this Sigma and some X, compute :math:`Sigma^{-1} * x`
+    def solve(self, X):
+        """Given this covariance and some X, compute :math:`Sigma^{-1} * x`
         """
         pass
 
-    @define_scope
-    def Sigma(self):
-        """return Sigma
-        """
-        return tf.matrix_inverse(self.Sigma_inv)
-
-    @define_scope
-    def Sigma_inv(self):
-        """ Sigma^{-1}. Override me with more efficient implementation in subclasses
-        """
-        return self.Sigma_inv_x(tf.diag(tf.ones([self.size],
-                                dtype=tf.float64)))
-
-    @define_scope
-    def logp(self):
-        """ Log-likelihood of this covariance (useful for regularization)
-        """
-        return tf.constant(0, dtype=tf.float64)
-
-
-class CovTFWrap(CovBase):
-    """ thin wrapper around a TF tensor
-    """
-    def __init__(self, Sigma):
-
-        self.L = tf.cholesky(Sigma)
-
-    def get_optimize_vars(self):
-        """ Returns a list of tf variables that need to get optimized to fit
-            this covariance
+    @property
+    def _prec(self):
+        """Expose the precision explicitly (mostly for testing /
+        visualization, materializing large covariances may be intractable)
         """
-        return []
+        return self.solve(tf.eye(self.size, dtype=tf.float64))
```

Comment:
I wonder if someone writes an inherent class and forgets to redefine `_prec`, will it take the default of inverse of this identity matrix without any warning?

Reply:
`self.solve` does `Sigma^-1 I` for the current Cov, and always has to be implemented (it's an abstract method). So basically, this is as-intended: you should rarely need to redefine `_cov` or `_prec`, since they're inefficient and mostly used for debugging, and they will always call the solve for you.

Reply:
(main current exception is in `CovAR1` where we redefine it, but that is because that's what you need for an efficient solve)